src/tasks/MonoTargetsTasks/MarshalingPInvokeScanner/MarshalingPInvokeScanner.cs

```diff
@@ -0,0 +1,333 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Reflection;
+using System.Reflection.PortableExecutable;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+internal enum Compatibility
+{
+    Compatible,
+    Incompatible,
+    Inconclusive
+}
+
+internal sealed class InconclusiveCompatibilityCollection
+{
+    private Dictionary<string, HashSet<string>> _data = new();
+
+    public bool IsEmpty => _data.Count == 0;
+
+    public void Add(string assyName, string namespaceName, string typeName)
+    {
+        HashSet<string>? incAssyTypes;
+
+        if(!_data.TryGetValue(assyName, out incAssyTypes))
+        {
+            incAssyTypes = new();
+            _data.Add(assyName, incAssyTypes);
+        }
+
+        incAssyTypes.Add(string.Join(":", namespaceName, typeName));
+    }
+
+    public HashSet<string> EnumerateForAssembly(string assyName)
+    {
+        if(_data.TryGetValue(assyName, out HashSet<string>? incAssyTypes))
+            return incAssyTypes!;
+
+        return new HashSet<string>();
+    }
+}
+
+internal sealed class MinimalMarshalingTypeCompatibilityProvider : ISignatureTypeProvider<Compatibility, object>
+{
+    // assembly name -> set of types needed for second pass
+    private InconclusiveCompatibilityCollection _inconclusive = new();
```

Comment:
I'm not 100% sure how to express this in SRM SignatureTypeProvider terms, but there's a nasty case in ECMA-335 having to do with enum types that are nested in a generic type:

```csharp
public class Outer<T> {
  public enum Inner {
    A, B, C
  }
}
```

([sharplab.io](https://sharplab.io/#v2:C4LglgNgPgsAUAYgHYFcIQIYCMIFMAEAJmAM7Z7zwACAzPlQEz4DyKwuATgDwAqAfPgDe8fPTq5UAW3wBJJEk5CRo/AEEANPgBCmgMLKAvvANA==))
In this case the enum type is generic:

```
.class nested public auto ansi sealed Inner<T>
        extends [System.Runtime]System.Enum
    {
        // Fields
        .field public specialname rtspecialname int32 value__
        .field public static literal valuetype Outer`1/Inner<!T> A = int32(0)
        .field public static literal valuetype Outer`1/Inner<!T> B = int32(1)
        .field public static literal valuetype Outer`1/Inner<!T> C = int32(2)

    } // end of class Inner
```

Do you know if this will be handled correctly here?

Reply:
The scanner does not handle your example well. I'll investigate further.